### ISSUE TITLE
How to find out why multiple nfs-ganesha head out of sync

### ISSUE THREAD

**Issue by panho66** [nfs-ganesha/nfs-ganesha#151]
Hi 
I have Gluster 3.8.9-1.el7 and nfs-ganesha 2.3.3-1.el7 installed in centos centos-release-7-3.1611.el7.centos.x86_64.

I got three nodes all have Gluster and nfs-ganesha. When a nfs client update a file another client connected to different nfs-ganesha and takes about 30s delay to see the change. Is UPCALL only work in nfs-ganesha 3.9 ?

here is my test
/bin/mount -t nfs4 10.161.100.48:/glusterV1 /mnt
/bin/mount -t nfs4 10.161.100.49:/glusterV1 /mnt2

[root@node2 ~]# date; cat  /mnt/hello.txt; cat /mnt2/hello.txt
Thu Mar  9 14:33:11 AEDT 2017
1234678
1234678

[root@node2 ~]# echo '1' > /mnt/hello.txt
[root@node2 ~]# date; cat  /mnt/hello.txt; cat /mnt2/hello.txt
Thu Mar  9 14:33:19 AEDT 2017
1
1234678
[root@node2 ~]# date; cat  /mnt/hello.txt; cat /mnt2/hello.txt
Thu Mar  9 14:34:08 AEDT 2017
1
1234678
[root@node2 ~]# date; cat  /mnt/hello.txt; cat /mnt2/hello.txt
Thu Mar  9 14:34:15 AEDT 2017
1
1

Gluster info
======================================
[root@node1log]# gluster vol info

Volume Name: glusterV1
Type: Replicate
Volume ID: 5e65f618-c81a-4727-8ab6-0dbfa8da1eb6
Status: Started
Snapshot Count: 0
Number of Bricks: 1 x 3 = 3
Transport-type: tcp
Bricks:
Brick1: node1:/bricks/brick1/glusterV1
Brick2: node2:/bricks/brick1/glusterV1
Brick3: node3:/bricks/brick1/glusterV1
Options Reconfigured:
server.root-squash: off
nfs.disable: on
performance.readdir-ahead: on
features.cache-invalidation: on
ganesha.enable: on
features.cache-invalidation-timeout: 600
performance.stat-prefetch: on
performance.md-cache-timeout: 60
cluster.enable-shared-storage: enable
nfs-ganesha: enable

Volume Name: gluster_shared_storage
Type: Replicate
Volume ID: 9e41e695-5a2f-4bff-8c0a-70631a45ecd6
Status: Started
Snapshot Count: 0
Number of Bricks: 1 x 3 = 3
Transport-type: tcp
Bricks:
Brick1: node1:/var/lib/glusterd/ss_brick
Brick2: node2:/var/lib/glusterd/ss_brick
Brick3: node3:/var/lib/glusterd/ss_brick
Options Reconfigured:
features.cache-invalidation: off
performance.readdir-ahead: on
nfs.disable: on
ganesha.enable: off
cluster.enable-shared-storage: enable
nfs-ganesha: enable

**Comment by soumyakoduri** [comment]
Upcall should work even on GLuster 3.8.x ..(though there were some issues in APIs being used. But AFAIR, @nixpanic applied work-around while building rpms on CENTOS). To check if upcalls are being sent, could you please collect packet trace and on the gluster node where NFS-Ganesha server is running and examine the packets for glusterfs.cbk program.

**Comment by panho66** [comment]
Hi soumyakoduri
I am not a linux person, how to collect package trace ? on network interface or dbus ?

Is this work ?
strace -p `pgrep -f 'dbus-daemon --system'` -s 3000 -ttt -xx -o foo.strace
cat foo.strace | grep glusterfs

**Comment by thotz** [comment]
Hi,
You can take the packet trace from node in which ganesha server is present using following command
tcpdump -w <filename to be saved> -i <interface name like eth0/em2 etc>  and not port 22

Upcall is supported and issue which happened mentioned by soumya koduri got resolved in branch in 3.8 (from 3.8.6 onwards).
But there are some ganesha related changes which is available in nfs-ganesha 2.4.2

Anyways please take the packet trace. So we can confirm whether upcall is send or not

**Comment by panho66** [comment]
the tcpdump file.
Test done by changing hello.txt from 3333 to 4444
[tcpdump.txt.gz](https://github.com/nfs-ganesha/nfs-ganesha/files/830600/tcpdump.txt.gz)

**Comment by thotz** [comment]
Hi,
You got the packet trace from ganesha server , right? I was not able to find packets related to upcall

**Comment by panho66** [comment]
yes, i have three nodes all have GlusterFS and Nfs-ganesha.
GlusterFS volume create with gluster peer probe <node1>, node2, node3 and
test done with nfs client connected to node1,. package capture created in node1.

So look like upcall not working... How could I diagnostic what causes it?

**Comment by panho66** [comment]
https://staged-gluster-docs.readthedocs.io/en/release3.7.0beta1/Features/upcall/
cache-invalidation is currently disabled by default. It can be enabled with the following command:
gluster volume set <volname> features.cache-invalidation on

I have features.cache-invalidation on but it seems UPCALL still not working.

https://soumyakoduri.wordpress.com/2015/02/25/glusterfs-understanding-upcall-infrastructure-and-cache-invalidation-support/

A new xlator(Upcall) has been defined to maintain and process state of the events which require server to send upcall notifications. For each I/O on a inode, we create/update a ‘upcall_inode_ctx’ and store/update the list of clients’ info ‘upcall_client_t’ in the context. The list is protected by mutex lock.

By default this xlator is OFF. It will be enabled by using a xlator option which will be added later.

How could I check UPCALL or xlator  is ON? How could I enable using  xlator option if it is OFF. Or I am looking at old information ?

One more question, if UPCALL is not enable, it is going to impact on NFS file lock lost or out of sync when one nfs-ganesh head shutdown and peacmaker move floating IP to another node ?

P.S. I am testing an application that heavily depends on NFS file lock to provide high available and I found  unnecessary application failover  due to NFS file lock lost.

**Comment by soumyakoduri** [comment]
I do not see any NFS write operations in the packet trace provided earlier. Could you please try following -
* First turn off glusterfs md-cache 
      gluster v set glusterV1 performance.stat-prefetch off
      gluster v set glusterV1 performance.md-cache-timeout 1
* Start pkt trace (tcpdump) on the node hosting 10.161.100.48 IP.
#tcpdump -w -i any and not port 22
* Now mount the volume using IPs of two different servers (10.161.100.48 & 10.161.100.49 servers ) and please try out the same operations which you had mentioned in the first comment on those fresh mount points. 
* <ctrl-c>/pkill tcpdump session
* please attach the pkt trace

**Comment by panho66** [comment]
[root@node2 ~]# /bin/mount -t nfs4 10.139.136.49:/glusterV1 /mnt2
[root@node2 ~]# /bin/mount -t nfs4 10.139.136.48:/glusterV1 /mnt
[root@node2 ~]# date; cat  /mnt/hello2.txt; cat /mnt2/hello2.txt
Tue Mar 14 09:53:55 AEDT 2017
333
333
[root@node2 ~]# echo '444' > /mnt/hello2.txt
[root@node2 ~]# date; cat  /mnt/hello2.txt; cat /mnt2/hello2.txt
Tue Mar 14 09:54:12 AEDT 2017
444
333
[root@node2 ~]# date; cat  /mnt/hello2.txt; cat /mnt2/hello2.txt
Tue Mar 14 09:54:41 AEDT 2017
444
333
[root@node2 ~]# date; cat  /mnt/hello2.txt; cat /mnt2/hello2.txt
Tue Mar 14 09:55:04 AEDT 2017
444
444
**This is odd and works for first time**
**[root@node2 ~]# echo '555' > /mnt2/hello2.txt
[root@node2 ~]# date; cat  /mnt/hello2.txt; cat /mnt2/hello2.txt
Tue Mar 14 09:55:15 AEDT 2017
555
555**
[root@node2 ~]# date; cat  /mnt/hello2.txt; cat /mnt2/hello2.txt
Tue Mar 14 09:55:19 AEDT 2017
555
555
[root@node2 ~]# echo '666' > /mnt2/hello2.txt
[root@node2 ~]# date; cat  /mnt/hello2.txt; cat /mnt2/hello2.txt
Tue Mar 14 09:55:27 AEDT 2017
555
666
[root@node2 ~]# date; cat  /mnt/hello2.txt; cat /mnt2/hello2.txt
Tue Mar 14 09:56:26 AEDT 2017
666
666

[10_139_136_48_tcpdump.dat.gz](https://github.com/nfs-ganesha/nfs-ganesha/files/839845/10_139_136_48_tcpdump.dat.gz)

[10_139_136_49_tcpdump.dat.gz](https://github.com/nfs-ganesha/nfs-ganesha/files/839846/10_139_136_49_tcpdump.dat.gz)

**Comment by soumyakoduri** [comment]
CACHE_INVALIDATION rpcs are captured in the packet trace ..that means server did send upcalls. Now we need to check if/why ganesha server couldn't process them.  Please run below cmd on the nodes where nfs-ganesha server is running and provide us the output file -

# gstack `pgrep ganesha` > ganesha_gstack

Also can you please have a sleep of 2-3 seconds post every write and then check the contents from each mount point.  Thanks!

**Comment by panho66** [comment]
I found that the dbus is really slow. Taking 3s to update 1 record
I ran two tests. First one causes client hang. Need to restart nfs-ganesha.
Here is the log message from second test from nfs client side

**The client script**
COUNTER=1
while true; do
  sleep 5
  echo "Counter=$COUNTER" > /mnt/hello.txt
  echo "$(date);$(cat /mnt/hello.txt); $(cat /mnt2/hello.txt);"
  COUNTER=$((COUNTER+1))
done

Counter=1 is from /mnt/hello.txt ( I think this is cache value from nfs-ganesha node1 but actual value in file system should be 286 same as node2 /mnt2/hello.txt
Counter=286 is from /mnt2/hello.txt  ( the value from first test. Still processing by nfs-ganesha)

Tue Mar 14 15:51:17 AEDT 2017;Counter=1; Counter=286;
Tue Mar 14 15:51:22 AEDT 2017;Counter=2; Counter=286;
Tue Mar 14 15:51:27 AEDT 2017;Counter=3; Counter=286;
Tue Mar 14 15:51:33 AEDT 2017;Counter=4; Counter=286;
Tue Mar 14 15:51:38 AEDT 2017;Counter=5; Counter=286;
Tue Mar 14 15:51:43 AEDT 2017;Counter=6; Counter=286;
Tue Mar 14 15:51:48 AEDT 2017;Counter=7; Counter=306;
Tue Mar 14 15:51:53 AEDT 2017;Counter=8; Counter=306;
Tue Mar 14 15:51:58 AEDT 2017;Counter=9; Counter=306;
Tue Mar 14 15:52:03 AEDT 2017;Counter=10; Counter=306;
Tue Mar 14 15:52:08 AEDT 2017;Counter=11; Counter=306;
Tue Mar 14 15:52:13 AEDT 2017;Counter=12; Counter=306;
Tue Mar 14 15:52:18 AEDT 2017;Counter=13; Counter=306;
Tue Mar 14 15:52:23 AEDT 2017;Counter=14; Counter=306;
Tue Mar 14 15:52:28 AEDT 2017;Counter=15; Counter=306;
Tue Mar 14 15:52:33 AEDT 2017;Counter=16; Counter=306;
Tue Mar 14 15:52:38 AEDT 2017;Counter=17; Counter=306;
Tue Mar 14 15:52:43 AEDT 2017;Counter=18; Counter=306;
Tue Mar 14 15:52:48 AEDT 2017;Counter=19; Counter=326;
Tue Mar 14 15:52:53 AEDT 2017;Counter=20; Counter=326;
Tue Mar 14 15:52:58 AEDT 2017;Counter=21; Counter=326;

**The script in nfs-ganesh node**
while true; do
     gstack 6712  >> ganesha_gstack.dat
    sleep 3
    echo "" >> ganesha_gstack.dat
    echo "$(date) sleep 3" >> ganesha_gstack.dat
    echo " $(cat /bricks/brick1/glusterV1/hello.txt) " >> ganesha_gstack.dat
    echo "" >> ganesha_gstack.dat
done

[ganesha.zip](https://github.com/nfs-ganesha/nfs-ganesha/files/840345/ganesha.zip)

**Comment by soumyakoduri** [comment]
Thats weird..  I tried out same test and it seem to be working fine -

Wed Mar 15 12:03:43 IST 2017;Counter=1; Counter=1;
Wed Mar 15 12:03:48 IST 2017;Counter=2; Counter=2;
Wed Mar 15 12:03:53 IST 2017;Counter=3; Counter=3;

But I used glusterfs-3.8.x and nfs-ganesha-2.4.x builds. There were some changes done to gfapi upcall APIs in glusterfs-3.8 latest releases. But the old API was still maintained in order not to break applications like nfs-ganesha-2.3.x. Probably there is a bug in there. I do not think these builds combination had been tested by anyone. Will it be possible for you to upgrade to nfs-ganesha-2.4.x and try out?

nfs-ganesha-2.4.x builds are available in https://buildlogs.centos.org/centos/7/storage/x86_64/gluster-3.9/

**Comment by panho66** [comment]
Good news, It works after nfs-ganesha upgraded to 2.4.1

here is what I got

glusterfs.x86_64                   3.8.10-1.el7                @centos-gluster38
glusterfs-api.x86_64               3.8.10-1.el7                @centos-gluster38
glusterfs-cli.x86_64               3.8.10-1.el7                @centos-gluster38
glusterfs-client-xlators.x86_64    3.8.10-1.el7                @centos-gluster38
glusterfs-fuse.x86_64              3.8.10-1.el7                @centos-gluster38
glusterfs-ganesha.x86_64           3.8.10-1.el7                @centos-gluster38
glusterfs-libs.x86_64              3.8.10-1.el7                @centos-gluster38
glusterfs-server.x86_64            3.8.10-1.el7                @centos-gluster38
libntirpc.x86_64                   1.4.3-1.el7                 installed
nfs-ganesha.x86_64                 2.4.1-1.el7                 installed
nfs-ganesha-gluster.x86_64         2.4.1-1.el7                 installed
userspace-rcu.x86_64               0.7.16-2.el7                @centos-gluster37

**This test  works**
cat ./testtest.sh
COUNTER=1
while true; do
sleep 5
echo "Counter=$COUNTER" > /mnt/hello.txt
echo "$(date);$(cat /mnt/hello.txt); $(cat /mnt2/hello.txt);"
COUNTER=$((COUNTER+1))
done

However, lock recovery seems broken.
I got this message when I did 'systemctl stop nfs-ganesha'. Once floating IP moved to pmc71dev1950, I got

16/03/2017 14:07:52 : epoch 58c9f365 : pmc71dev1950.auiag.corp : ganesha.nfsd-450[dbus_heartbeat] nfs4_start_grace :STATE :WARN :NFS Server GRACE duration should at least match LEASE period. Current configured values are GRACE(5), LEASE(120)
16/03/2017 14:07:52 : epoch 58c9f365 : pmc71dev1950.auiag.corp : ganesha.nfsd-450[dbus_heartbeat] nfs4_start_grace :STATE :EVENT :NFS Server Now IN GRACE, duration 5
16/03/2017 14:07:52 : epoch 58c9f365 : pmc71dev1950.auiag.corp : ganesha.nfsd-450[dbus_heartbeat] nfs4_start_grace :STATE :EVENT :NFS Server recovery event 5 nodeid -1 ip pmc71dev1949.auiag.corp
16/03/2017 14:07:52 : epoch 58c9f365 : pmc71dev1950.auiag.corp : ganesha.nfsd-450[dbus_heartbeat] nfs4_load_recov_clids_nolock :CLIENT ID :EVENT :Recovery for nodeid -1 dir (/var/lib/nfs/ganesha/pmc71dev1949.auiag.corp/v4recov)

and my application failover to standby because of lost NFS file lock.

So, the question is lock recovery work or not and how could I check ?

P.S. I have this and not expect lock file lease expired soon

in /etc/ganesha/exports/export.glusterV1.conf
NFSV4 {
   Lease_Lifetime = 120;
   Grace_Period = 5;
}

**Comment by soumyakoduri** [comment]
Glad nfs-ganesha-2.4 worked. 

Wrt other issues mentioned -

>>> 
16/03/2017 14:07:52 : epoch 58c9f365 : pmc71dev1950.auiag.corp : ganesha.nfsd-450[dbus_heartbeat] nfs4_start_grace :STATE :WARN :NFS Server GRACE duration should at least match LEASE period. Current configured values are GRACE(5), LEASE(120)
16/03/2017 14:07:52 : epoch 58c9f365 : pmc71dev1950.auiag.corp : ganesha.nfsd-450[dbus_heartbeat] nfs4_start_grace :STATE :EVENT :NFS Server Now IN GRACE, duration 5
16/03/2017 14:07:52 : epoch 58c9f365 : pmc71dev1950.auiag.corp : ganesha.nfsd-
<<<

[Soumya] When a IP failover happens, we put all the servers into grace so that clients can reclaim their lost state. And it is recommended that for clients to reclaim their state, server grace period should at-least match lease_period. Hence the above warning is displayed. Is there any reason for having 'Lease_Lifetime' configured to much higher value compared to 'Grace_Period' value?

>>>
450[dbus_heartbeat] nfs4_start_grace :STATE :EVENT :NFS Server recovery event 5 nodeid -1 ip pmc71dev1949.auiag.corp
16/03/2017 14:07:52 : epoch 58c9f365 : pmc71dev1950.auiag.corp : ganesha.nfsd-450[dbus_heartbeat] nfs4_load_recov_clids_nolock :CLIENT ID :EVENT :Recovery for nodeid -1 dir (/var/lib/nfs/ganesha/pmc71dev1949.auiag.corp/v4recov)
<<<
[Soumya] Above messages are just event messages, nothing to be concerned about. This code path needs to be cleaned up as we need not use nodeid (like in case of Gluster) for clustered solution.

So to summarize, there are no failures in lock recovery logged but just a warning stating that Grace should be >= Lease time.

>>>
and my application failover to standby because of lost NFS file lock.
<<<
[Soumya] This is a different issue and not related to lock-recovery messages above. Application went to 'standby' - does it mean floating IPs went to STOPPED state? Could you please check that using 'pcs status' command?

Also please provide the output of 'pcs property'.

**Comment by panho66** [comment]
>>>Is there any reason for having 'Lease_Lifetime' configured to much higher value compared to 'Grace_Period' value?

We have network issue couple times a year which may last for 60s. The application I try to get it working with nfs-ganesha has Active and StandBy both running at the same time. The StandBy try to take over lock file and there is no turning parameter or wait time as long as lock file acquired StandBy concluded that Active is no longer available and StandBy starts up. Once StandBy starting up, Active lost lock file even it is OK Active is going to shutdown.

Set Lease_Lifetime to 120s to minimize outage caused by Network issue. (I hope :) ).
Set Grace_Period=> Lease_Lifetime, it causes unnecessary application failover if one of the Gluster node or nfs-ganesha failed. Lock file will be expired and StandBy is going to take over if Grace_Period=Lease_Lifetime

After I set Grace_Period to 25, it works perfectly. There is no unnecessary application failover when I stopped nfs-ganesha. I think I read somewhere about pacemaker floating IP takes about 20 to 22s to failover.

But it is sad that the maximum value of Lease_Lifetime which can be set in nfs-ganesha is 120s....

**Comment by ffilz** [comment]
You don't want grace to be less than lease, if it is, clients may not detect server crash in time to reclaim lost locks during grace period.

That said, I don't know if there's any solid reason to really limit lease time or grace period...

**Comment by panho66** [comment]
>>>You don't want grace to be less than lease, if it is, clients may not detect server crash in time to reclaim lost locks during grace period.
The client in this case has active and standby watcher, standby watcher pulling lock if it can and before active could reclaimed lock then standby component concluded that active component no longer available. Standby component starts up and becomes active. My observation during the test by shutdown nfs-ganesha,  the  Client do not do reclaim lost locks or grace period too short or steal by standby component in race condition. I need to raise support call to the product to confirm what client actually doing in the case

Correct me if I am wrong, Grace Period is useful if NFS servers having active and standby node. when active node have failed, standby node started up. Grace period is required for client to help nfs server to recovery lock by reclaim lock file which is only allow in grace period.

However, nfs-ganesha is active/active. When one node failed, floating IP move to anther node where nfs-ganesha is also active.

So, the questions are:
1) nfs-ganesha needs client help using reclaim lock during grace period to recover lock information or not ?
2) the second nfs-ganesha node should have lock information from client before first nfs-ganesha node failed. How up to date information is?
2) how second node got the lock information? via glusterfs shared storage? and UPCALL ? That imply it is sync in millisecond ?
3)  When failed node back online and nfs-ganesha started, where it got lock information? via glusterfs shared storage? and how long it takes to in sync with other node to have latest lock information ?

**Comment by ffilz** [comment]
Closing this due to age, feel free to re-open if it's still of interest and you have explored using a currently supported version of Ganesha (V2.8 or V3 at this time).
